onnx/test/symbolic_shape_test.py

```diff
@@ -0,0 +1,135 @@
+        # the symbolic shape of C and output should be the same
+        assert self._get_shape_from_name(inferred_model, 'C') == self._get_shape_from_name(inferred_model, 'output')
+
+    def test_two_symbolic_concat(self):  # type: () -> None
+        concat1 = helper.make_node('Concat', inputs=['A', 'B'], outputs=['C'], name='Concat', axis=1)
+        concat2 = helper.make_node('Concat', inputs=['C', 'D'], outputs=['E'], name='Concat', axis=1)
+        cast = onnx.helper.make_node('Cast',
+            inputs=['E'],
+            outputs=['output'],
+            to=getattr(TensorProto, 'FLOAT'))
+        graph_def = helper.make_graph(name='test_graph',
+            nodes=[concat1, concat2, cast],
+            inputs=[helper.make_tensor_value_info('A', TensorProto.FLOAT, [2, 'A']),
+                helper.make_tensor_value_info('B', TensorProto.FLOAT, [2, 3]),
+                helper.make_tensor_value_info('D', TensorProto.FLOAT, [2, 'D'])],
+            outputs=[helper.make_tensor_value_info('output', TensorProto.FLOAT, [2, 'M'])]
+        )
+
+        onnx_model = make_model(graph_def)
+        inferred_model = onnx.shape_inference.infer_shapes(onnx_model, strict_mode=True)
+        self._assert_valueinfo_shape(inferred_model, [
+            make_tensor_value_info("C", TensorProto.FLOAT, (2, -1)),
+            make_tensor_value_info("E", TensorProto.FLOAT, (2, -1))])
+        # the symbolic shape of E and output should be the same
+        assert self._get_shape_from_name(inferred_model, 'E') == self._get_shape_from_name(inferred_model, 'output')
+
+    def test_duplicate_symbolic_shape(self):  # type: () -> None
+        concat1 = helper.make_node('Concat', inputs=['A', 'B'], outputs=['C'], name='Concat', axis=1)
+        concat2 = helper.make_node('Concat', inputs=['C', 'D'], outputs=['E'], name='Concat', axis=1)
+        cast = onnx.helper.make_node('Cast',
+            inputs=['E'],
+            outputs=['output'],
+            to=getattr(TensorProto, 'FLOAT'))
+        graph_def = helper.make_graph(name='test_graph',
+            nodes=[concat1, concat2, cast],
+            inputs=[helper.make_tensor_value_info('A', TensorProto.FLOAT, [2, 'unk__0']),
+                helper.make_tensor_value_info('B', TensorProto.FLOAT, [2, 3]),
+                helper.make_tensor_value_info('D', TensorProto.FLOAT, [2, 'D'])],
+            outputs=[helper.make_tensor_value_info('output', TensorProto.FLOAT, [2, 'unk__1'])]
+        )
+
+        onnx_model = make_model(graph_def)
+        original_count = self._count_unique_dim_param_number(onnx_model)
+        inferred_model = onnx.shape_inference.infer_shapes(onnx_model, strict_mode=True)
+        inferred_count = self._count_unique_dim_param_number(inferred_model)
+        # new symbol 'unk__2' should be generated to prevent duplicate so the count will be count + 1
+        assert inferred_count == original_count + 1, '%s%s' % (inferred_model, onnx_model)
```

Comment:
shouldn't this be original_count + 2?

Concatenation of A[2, 'unk__0'], B[2, 3] should result in output C of shape [2, 'unk__2'] 
Concatenation of C[2, 'unk__2'] , D[2, 'D'] should result in output of shape [2, 'unk__3']
what am I missing?

Reply:
Good catch. I think my previous example here is not a proper one (looks confusing):
- The original symbol table is {'unk__0', 'unk__1', 'D'}
- After inference symbol table is {'unk__2', 'unk__0', 'unk__3', 'D'}

Although original output shape is 'unk__1', but the inferred output shape will be 'unk__3', which replaces original 'unk__1'. 

To resolve confusion, I will use 'unk__0' as output shape. Then the inferred count will be count + 2 as you said.